fix: allow to use the new provider.iam.role.statements

t-richard opened the pull request:
Should fix https://github.com/getlift/lift/issues/47

Here we add to `provider.iam.role.statements` if it exists or add to `provider.iamRoleStatements` if not (so that we are still backward-compatible with older versions of the serverless framework.

TODO
* [x] Test it 
* [x] Fix the type issue
<!--
Please explain the motivation behind the changes.

In other words, explain **WHY** instead of **WHAT**.

If you are adding a new config option in a component, please explain the use case with an example.
-->

t-richard commented:
Not sure how to fix the CI. 

> Property 'statements' does not exist on type 'AwsLambdaRole | { name?: string | undefined; managedPolicies?: AwsArn[] | undefined; statements?: AwsIamPolicyStatements | undefined; permissionBoundary?: string | undefined; tags?: AwsResourceTags | undefined; }'
Property 'statements' does not exist on type 'string'.

It seems like the type definition is correct and that `statements` should match it. Not sure if the problem comes from the definition, the `tsc` checks or the code.

Let me know if you have any idea on how to fix.

mnapoli commented:
Thanks! The change itself looks really good. TypeScript doesn't approve of using/changing anything that isn't specified in the types (even if you double-check the property exists before).

I can take care of fixing the types on Monday! (and maybe add some tests too)

t-richard commented:
@mnapoli You're welcome.

I've fixed the type issue: basically, role can be a string, a cfn intrisinc expression or an object (which can contain `statements`). Ensuring we have an object filters out the string and checking that `statements` exists filters ou the intrisinc function and that it's not undefined. This way, Typescript is happy :smile: 

I've also added some tests but it's not covering all cases for now. If you think this is heading the right way, I'll check that again tomorrow.

mnapoli commented on test/unit/permissions.test.ts:
```diff
@@ -0,0 +1,62 @@
+import { pluginConfigExt, runServerless } from "../utils/runServerless";
+
+describe("permissions", () => {
+    it("should not override user-defined role", async () => {
```
👌

mnapoli commented on test/unit/permissions.test.ts:
```diff
@@ -0,0 +1,62 @@
+                provider: {
+                    iam: {
+                        role: {
+                            statements: [
```
The test name mentions `iamRoleStatements` but actually uses `iam.role.statements`, is this a mistake?

t-richard replied:
yes it is :sweat_smile:

mnapoli commented:
> This way, Typescript is happy 😄

Perfect

> I've also added some tests but it's not covering all cases for now. If you think this is heading the right way, I'll check that again tomorrow.

And perfect again, yes those tests look good! I see you started that but ideally we should have:

- a test that covers `iamRoleStatements`
- a test that covers `iam.role.statements`

fredericbarthelet commented on src/plugin.ts:
```diff
@@ -199,6 +199,13 @@ class LiftPlugin {
+        const role = this.serverless.service.provider.iam?.role;
+
+        if (typeof role === "object" && "statements" in role) {
+            role.statements?.push(...statements);
```
shouldn't you `return` here as well to prevent creation of `this.serverless.service.provider.iamRoleStatements` in case version of sls is up to date ?

t-richard replied:
indeed, did it in my first commit but forgot it when refactoring to fix the type issue :+1:

t-richard commented:
Fixed the review comments and added some more tests to test with the new syntax, the old syntax and when no statements are provided.

mnapoli commented:
Thank you!
